samples/rust/rust_thread.rs

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rust thread APIs sample
+
+#![no_std]
+#![feature(allocator_api, global_asm)]
+
+use alloc::boxed::Box;
+use core::sync::atomic::{AtomicBool, Ordering};
+use kernel::prelude::*;
+use kernel::{
+    c_str, pr_cont,
+    sync::Ref,
+    thread::{schedule, Thread},
+};
+
+module! {
+    type: RustThread,
+    name: b"rust_thread",
+    author: b"Rust for Linux Contributors",
+    description: b"Rust thread APIs sample",
+    license: b"GPL v2",
+    params: {
+    },
+}
+
+struct RustThread;
+
+impl KernelModule for RustThread {
+    fn init() -> Result<Self> {
+        pr_info!("Rust thread APIs sample (init)");
+
+        // Test threads.
+        {
+            Thread::try_new(c_str!("rust-thread0"), || {
+                pr_info!("Hello Rust Thread");
+                Ok(())
+            })?
```

Comment:
Looking at this, the following thought came to mind: what if we never call `wake_up` (either on purpose or by omission, i.e., not having realised that it needs to called)?

The thread and its state would be stuck in no-man's land, right? I think we should make it harder for such a state to be reached; another way to put it: a better design would one in which this no-man's state is reached only when the user explicitly intends it.

Two things come to mind:
1.  We should offer a constructor that automatically starts the thread in addition to the one we have now. In fact, I think the `try_new` one should automatically wake up the thread and the other one could be something like `try_new_paused` (or something to that effect).

2. Do thread creation in two stages with different types. The first stage is one in which the thread was never started; there are three ways to come out of it (they would all consume this initial type and optionally return a new one): starting the thread (returns the `Thread` as you have here); stopping the thread (doesn't return anything); upgrading (doesn't start the thread, but returns `Thread` as you have here). If none of these actions are taken, i.e., we default to stop.

Some examples to illustrate what I mean:

```rust
Thread::try_new(...); // automatically runs the closure

Thread::try_new_paused(...).run(); // automatically runs the closure

// Creates the thread but never runs it and automatically cleans
// everything up (no accidental leaks).
Thread::try_new_paused();

// Same as above
Thread::try_new_paused().stop();

// `upgrade` is not the best name, we should think of something better.
//
// creates the thread, never runs it, but don't do cleanups beyond 
// `put_task_struct`, which is basically what we have today but with
// explicit action from the user (calling `upgrade`).
Thread::try_new_paused().upgrade(); 
```

To implement the above, `Thread::try_new_paused` would return a different type, something like `InitialThread` that would have the `run`, `upgrade`, `stop` methods plus a `Drop` implementations that calls `stop` (and would be inhibited by the calls to `run` and `upgrade`).

WDYT?

Reply:
Two more examples to make the role of `InitialThread` clearer:

```rust
fn example1() -> Result {
    let t = Thread::try_new_paused(...);
    // Do other stuff. If it fails, and we return, `t` never runs and is cleaned up.
    let t = t.run();
    // Now we can use `t` as a `Thread`.
}

fn example1() -> Result {
    let t = Thread::try_new_paused(...).upgrade();
    // Do other stuff. If it fails, and we return, `t` never runs and leaks memory.
    let t = t.run();
    // Now we can use `t` as a `Thread`.
}

Reply:
> Looking at this, the following thought came to mind: what if we never call `wake_up` (either on purpose or by omission, i.e., not having realised that it needs to called)?
> 

The `Thread::drop` will handle it in that case. So if you do
```rust
fn do_something() {
    let t = Thread::try_new(...)?;
    // never call t.wake_up();
}
``` 
the `Thread` will be freed when `do_something()` returns. IOW, the lifetime of a kernel thread is the same as the lifetime of a `Thread`. For a (kernel) thread that forever runs, kernel must keep a reference to is somewhere, otherwise kernel won't have a way to manipulate it.

> The thread and its state would be stuck in no-man's land, right? I think we should make it harder for such a state to be reached; another way to put it: a better design would one in which this no-man's state is reached only when the user explicitly intends it.
> 

The interface that I want to provide to users is that if a `Thread` is alive, then the thread must be there. And if the users want to keep a thread alive for a long time (maybe forever), they are required to manage the lifetime of `Thread`.

Reply:
> The `Thread::drop` will handle it in that case. So if you do
> 
> ```rust
> fn do_something() {
>     let t = Thread::try_new(...)?;
>     // never call t.wake_up();
> }
> ```
> 
> the `Thread` will be freed when `do_something()` returns. IOW, the lifetime of a kernel thread is the same as the lifetime of a `Thread`. For a (kernel) thread that forever runs, kernel must keep a reference to is somewhere, otherwise kernel won't have a way to manipulate it.

If we never call `t.wake_up`, `do_something()` will never run and therefore never return, right? (Or am I missing something here?)

IOW, it looks to me that we're leaking by inaction (i.e., not calling `t.wake_up`). I'm saying that we should design things to avoid accidental leaks like this one. If users actually do want to leak, they should do it explicitly (e.g., by calling some function that will result in the leak).

Reply:
> > The `Thread::drop` will handle it in that case. So if you do
> > ```rust
> > fn do_something() {
> >     let t = Thread::try_new(...)?;
> >     // never call t.wake_up();
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > the `Thread` will be freed when `do_something()` returns. IOW, the lifetime of a kernel thread is the same as the lifetime of a `Thread`. For a (kernel) thread that forever runs, kernel must keep a reference to is somewhere, otherwise kernel won't have a way to manipulate it.
> 
> If we never call `t.wake_up`, `do_something()` will never run and therefore never return, right? (Or am I missing something here?)
> 

No, please see the comments below (I adjust a little on the return type of `do_something()`.

```rust
fn do_something() -> Result {
     let t = Thread::try_new(...)?; // create a thread, and thread doesn't start to run
     // if t.wake_up() here, the thread will start running
     return Ok(()); // here `drop(t)` will be called, if `t` hasn't started, `t` will get recycled, otherwise, we wait `t` to finish and then recycle it.
}

```
> IOW, it looks to me that we're leaking by inaction (i.e., not calling `t.wake_up`). I'm saying that we should design things to avoid accidental leaks like this one. If users actually do want to leak, they should do it explicitly (e.g., by calling some function that will result in the leak).

I don't think we leak anything here, all thread related resources are bound to `Thread`, and they will get recycled when `Thread` reach the end of life, that's how I want to design the API.

Reply:
I re-read your follow-up comments, maybe I was missing your point ;-(

Looks like your point is that we avoid as hard as possible the case where a user call `try_new()` but forget to call `wake_up()`, because in that case, the thread related resources are allocated, but no one actually uses it. Am I getting it right?

Reply:
> I re-read your follow-up comments, maybe I was missing your point ;-(
> 
> Looks like your point is that we avoid as hard as possible the case where a user call `try_new()` but forget to call `wake_up()`, because in that case, the thread related resources are allocated, but no one actually uses it. Am I getting it right?

Yes, that's what I'm trying to say. 

[And thanks for the extra effort in trying to decipher what I'm saying, it's much appreciated :) ]

Reply:
Sorry for late response, and glad I get your point.

I don't think that's an issue personally, because we have `kthread_create()` and `kthread_run()` API in kernel, and kernel developers are familiar with the concept that a created kthread needs a `wake_up()` to run.

But it seems I'm missing the `Thread::drop()` implementation in this PR, let me add it and update the PR to see I can resolve your concerns ;-)

Reply:
@wedsonaf I understand your concern and I agree that's a problem we should resolve, however, I'd rather keep the `Thread` as simple as possible, and want to solve that issue with higher level abstraction. Reasons as below:

1. In my original design, `Thread` is the counterpart of `JoinHandle` in Rust std, it provides a way to wait for the thread to finish and get the return value of the thread function. This also means `Thread` shares the same semantics for `drop` as `JoinHandle`:

> A JoinHandle detaches the associated thread when it is dropped, which means that there is no longer any handle to thread and no way to join on it. (from https://doc.rust-lang.org/std/thread/struct.JoinHandle.html)

IOW, if an owner for `Thread` was to make sure a thread reaches to the end of life, the only way should be calling `Thread::stop`.

2. One difference between `Thread` and `JoinHandle` is that `Thread` can represent a thread that doesn't start yet, we need this ability because sometimes the kernel developers need to do some initialization work between thread creation and thread running, e.g. setting a global variable with the thread pointer (may not needed for Rust) or binding the thread to a particular CPU. This leads us to your suggestion on two types for threads. I like your suggestion, but I'm not sure that's best option for users, an alternative solution in my mind is function like:

```rust
impl Thread {
    pub fn try_init_and_run<F, I>(f: F, init: I, ...) -> Result<Self> 
    where F: FnOnce(),
          I : FnOnce(Deref<Target=task>) {
    }
}

```

IOW, we provide an init closure so that every thread created must be started.

3. Even every thread is in started state after creation, we are still subject to resource waste if the thread function has a dead loop in it (or wait for some conditions that cannot be satisfied), which makes no difference than creating a thread but forgetting to start or stop it. It's a user error (or mistake) and it's "not a memory safety issue(TM)" ;-). So even if we have an extra type for not-started threads, we still have other similar user errors. Therefore I'm in favor of keeping `Thread` as it is (of course, your suggestion about making it a `Deref<Task>` makes sense to me). `Thread` is by design the lowest layer to interact with C kernel thread API, so anyone using it must take some caution. We can always provide a high-level layer if some user pattern is frequently used and well defined.

So I'm going to keep `Thread` as it is, and try to move as many functions to `Task` as possible, resolve other comments from your and send an updated version. Thoughts?

Reply:
> But it seems I'm missing the `Thread::drop()` implementation in this PR, let me add it and update the PR to see I can resolve your concerns ;-)

What is the `drop` implementation going to do?

> So I'm going to keep `Thread` as it is, and try to move as many functions to `Task` as possible, resolve other comments from your and send an updated version. Thoughts?

Let's forget my suggestion for a bit. Here's a simpler way to phrase the problem, suppose we have this code:

```rust
fn do_something() -> Result {
     let t = Thread::try_new(...)?;
     init_some_state()?; // <- This fails
     t.wake_up();
     return Ok(());
}
```

Now suppose `init_some_state()` fails and we return without calling `t.wake_up()`. We're leaking memory. But part of what we claim with this project is that we simplify error handling by automatically cleaning up resources on error.

How do you feel about this? I'd feel much better if we didn't leak in the case above. I'm neutral on _how_ we address it.

Reply:
> > But it seems I'm missing the `Thread::drop()` implementation in this PR, let me add it and update the PR to see I can resolve your concerns ;-)
> 
> What is the `drop` implementation going to do?
> 

The implementation of `drop` would be simply `put_task_struct`, and yes, I know it cannot resolve the "leak" in your case.

> > So I'm going to keep `Thread` as it is, and try to move as many functions to `Task` as possible, resolve other comments from your and send an updated version. Thoughts?
> 
> Let's forget my suggestion for a bit. Here's a simpler way to phrase the problem, suppose we have this code:
> 
> ```rust
> fn do_something() -> Result {
>      let t = Thread::try_new(...)?;
>      init_some_state()?; // <- This fails
>      t.wake_up();
>      return Ok(());
> }
> ```
> 
> Now suppose `init_some_state()` fails and we return without calling `t.wake_up()`. We're leaking memory. But part of what we claim with this project is that we simplify error handling by automatically cleaning up resources on error.
> 

My point is that even if we call `t.wake_up()`, we could still "leak" memory, 

```rust
fn do_something() -> Result {
      let t = Thread::try_new(|| { wait_for_init(...); ... })?;
      t.wake_up();

      init_some_state()?; // <- This is supposed to unblock the wait_for_init, but it fails
      return Ok(());
}
```

, because although `t` is started, but it will never run. Similar if you change the `try_new` to `try_new_and_run`.

> How do you feel about this? I'd feel much better if we didn't leak in the case above. I'm neutral on _how_ we address it.

I agree that we should resolve this, but I'm not sure `Thread` is the proper layer that we resolve it. The current design of `Thread` is combination of `JoinHandle` in Rust std and Linux C kernel API, both of which are already known and somehow well-defined, but what's we don't know is the usage pattern that Rust users may choose to manage the kernel threads, for example, will Rust user prefer one-shot threads (only run one time and return) or long-live threads? Given the uncertainness, I'm not sure the best API the users want. The current rule of `Thread` is easy, if one wants to make sure all the resources of a thread is recycled, one must call `stop`, otherwise, there is no guarantee. Upon the `Thread`, we can build other layers to resolve various leaking issues.

Reply:
> My point is that even if we call `t.wake_up()`, we could still "leak" memory,

Sure, but I think that the fact that we _may_ 'leak' later has no bearing on the fact that we _definitely_ leak between `try_new` and `wake_up`.

> I agree that we should resolve this, but I'm not sure `Thread` is the proper layer that we resolve it. The current design of `Thread` is combination of `JoinHandle` in Rust std and Linux C kernel API, both of which are already known and somehow well-defined, but what's we don't know is the usage pattern that Rust users may choose to manage the kernel threads, for example, will Rust user prefer one-shot threads (only run one time and return) or long-live threads? Given the uncertainness, I'm not sure the best API the users want. The current rule of `Thread` is easy, if one wants to make sure all the resources of a thread is recycled, one must call `stop`, otherwise, there is no guarantee. Upon the `Thread`, we can build other layers to resolve various leaking issues.

It seems to me that part of your argument is that the situation above is unique. So let me show 3 examples from the project that, in my view, show a pattern that I think we should follow here:

1. Reference-counted object:
```rust
    let r = UniqueRef::try_new(...)?;
    do_something()?; // <- if this fails, we automatically free the object.
    r.into() // <- converts UniqueRef<T> to Ref<T> -- users may leak this by creating
             // circular references and many other ways, but we still clean up if
             // do_something fails.
```

2. Red-black tree nodes:
```rust
    let r = RBTree::try_reserve_node()?;
    do_something()?; // <- if this fails, we automatically free the node reservation.
    tree.insert(r.into_node()); // <- this converts a reservation into a node and inserts it
                                // into the tree -- users may "leak" it by never removing it from the
                                // tree but we still clean up the reservation if do_something above 
                                // fails. 
```

3. File descriptor:
```rust
    let r = FileDescriptorReservation::new(...)?;
    do_something()?; // <- if this fails, we automatically free the fd reservation.
    r.commit(file); // <- ties this fd reservation to the given file -- users may "leak" the
                    // file descriptors by never closing it, but we still clean up the reservation
                    // if do_something fails.
```

So perhaps one way to address your concern wrt to keeping things simple would be something like this: `Thread::try_new()` creates and automatically runs a thread -- I suspect this would be usual way of doing things, and would be similar to userspace and `kthread_run`.

For the more unusual scenario where one needs to do some work before letting the thread run, we'd have a type 'PausedThread' (or something similar, but would play the same role as `UniqueRef`, `FdReservation` and `NodeReservation` in the examples above). `PausedThread::try_new` creates a paused thread that is automatically cleaned up if it goes out of scope. Eventually the users would call `paused_thread.into_thread()` which would consume the paused thread and turn it into a regular `Thread` by just waking it up. (We may even avoid the extra reference count increment on the new task if we have `PausedThread::wake_up` that consumes the `PausedThread` but doesn't return anything -- we don't need to increment the refcount because we're giving up any pointers to the new task).

So we keep `Thread` as simple as it gets. The extra 'complexity' only comes in when users want the less common scenarios, and even then it is contained in `PausedThread`. And we never leak paused threads if an error occurs before they run.

Reply:
> > My point is that even if we call `t.wake_up()`, we could still "leak" memory,
> 
> Sure, but I think that the fact that we _may_ 'leak' later has no bearing on the fact that we _definitely_ leak between `try_new` and `wake_up`.
> 
> > I agree that we should resolve this, but I'm not sure `Thread` is the proper layer that we resolve it. The current design of `Thread` is combination of `JoinHandle` in Rust std and Linux C kernel API, both of which are already known and somehow well-defined, but what's we don't know is the usage pattern that Rust users may choose to manage the kernel threads, for example, will Rust user prefer one-shot threads (only run one time and return) or long-live threads? Given the uncertainness, I'm not sure the best API the users want. The current rule of `Thread` is easy, if one wants to make sure all the resources of a thread is recycled, one must call `stop`, otherwise, there is no guarantee. Upon the `Thread`, we can build other layers to resolve various leaking issues.
> 
> It seems to me that part of your argument is that the situation above is unique. So let me show 3 examples from the project that, in my view, show a pattern that I think we should follow here:
> 
> 1. Reference-counted object:
> 
> ```rust
>     let r = UniqueRef::try_new(...)?;
>     do_something()?; // <- if this fails, we automatically free the object.
>     r.into() // <- converts UniqueRef<T> to Ref<T> -- users may leak this by creating
>              // circular references and many other ways, but we still clean up if
>              // do_something fails.
> ```
> 
> 2. Red-black tree nodes:
> 
> ```rust
>     let r = RBTree::try_reserve_node()?;
>     do_something()?; // <- if this fails, we automatically free the node reservation.
>     tree.insert(r.into_node()); // <- this converts a reservation into a node and inserts it
>                                 // into the tree -- users may "leak" it by never removing it from the
>                                 // tree but we still clean up the reservation if do_something above 
>                                 // fails. 
> ```
> 
> 3. File descriptor:
> 
> ```rust
>     let r = FileDescriptorReservation::new(...)?;
>     do_something()?; // <- if this fails, we automatically free the fd reservation.
>     r.commit(file); // <- ties this fd reservation to the given file -- users may "leak" the
>                     // file descriptors by never closing it, but we still clean up the reservation
>                     // if do_something fails.
> ```
> 

It makes more sense considering this is a widely used pattern ;-)

> So perhaps one way to address your concern wrt to keeping things simple would be something like this: `Thread::try_new()` creates and automatically runs a thread -- I suspect this would be usual way of doing things, and would be similar to userspace and `kthread_run`.
> 

I probably name it as `Thread::try_run()`.

> For the more unusual scenario where one needs to do some work before letting the thread run, we'd have a type 'PausedThread' (or something similar, but would play the same role as `UniqueRef`, `FdReservation` and `NodeReservation` in the examples above). `PausedThread::try_new` creates a paused thread that is automatically cleaned up if it goes out of scope. Eventually the users would call `paused_thread.into_thread()` which would consume the paused thread and turn it into a regular `Thread` by just waking it up. (We may even avoid the extra reference count increment on the new task if we have `PausedThread::wake_up` that consumes the `PausedThread` but doesn't return anything -- we don't need to increment the refcount because we're giving up any pointers to the new task).
> 
> So we keep `Thread` as simple as it gets. The extra 'complexity' only comes in when users want the less common scenarios, and even then it is contained in `PausedThread`. And we never leak paused threads if an error occurs before they run.

OK, I will update this PR accordingly.

Reply:
> So we keep Thread as simple as it gets. The extra 'complexity' only comes in when users want the less common scenarios, and even then it is contained in PausedThread. And we never leak paused threads if an error occurs before they run. 

@wedsonaf there is an interesting limitation: should `PausedThread` impl `Deref<Target=Task>`? It makes sense it does, because most of operations between `try_new` and `wake_up` are task related: e.g. set nice value. However, this conflicts with having a `Task::wake_up()`, because directly calling `wake_up` on `PausedThread` violates the type invariants. But having a `Task::wake_up()` is quite natural, which I already did in this PR. Any idea?

Reply:
That's indeed problematic. I agree that we should have `Task::wake_up`, especially because it's used/useful in other scenarios, but that seems to prevent us from implementing `Deref<Target = Task>` for `PausedThread`. I can think of two ways of going about this, none of them ideal:
1. Simply not implement `Deref<Target = Task>`, which means that we need to implement a bunch of `Task` methods in `PausedThread` and forward them.
2. Implement an unsafe variant of `Deref` (e.g., an `as_task_unsafe()` method) with the safety requirement being that the caller must not call `wake_up` on the returned task.

Or perhaps we can do a combination of both, for example, we only implement the most common `Task` methods in `PausedThread` (e.g., `nice()`). Other users can go the `unsafe` route of caling `as_task_unsafe()` or simply adding the method to `PauseThread` if it is common enough. WDYT?

Reply:
> That's indeed problematic. I agree that we should have `Task::wake_up`, especially because it's used/useful in other scenarios, but that seems to prevent us from implementing `Deref<Target = Task>` for `PausedThread`. I can think of two ways of going about this, none of them ideal:
> 
> 1. Simply not implement `Deref<Target = Task>`, which means that we need to implement a bunch of `Task` methods in `PausedThread` and forward them.
> 2. Implement an unsafe variant of `Deref` (e.g., an `as_task_unsafe()` method) with the safety requirement being that the caller must not call `wake_up` on the returned task.
> 
> Or perhaps we can do a combination of both, for example, we only implement the most common `Task` methods in `PausedThread` (e.g., `nice()`). Other users can go the `unsafe` route of caling `as_task_unsafe()` or simply adding the method to `PauseThread` if it is common enough. WDYT?

How about a `AutoStopThread` instead of `PausedThread`? `AutoStopThread` impl `Deref<Target=Task>` and can convert to or **from** `Thread`, which means `AutoStopThread` can represent a thread that is already running, the only difference is that `AutoStopThread` will call `kthread_stop` when it goes out of scope (i.e. in `AutoStopThread::drop`) while `Thread::drop` is just `put_task_struct`.

`AutoStopThread` does the same thing as `PausedThread`: cleaning up when `drop`. The only downside is that we need to `get_task_struct` when we create a new `AutoStopThread` for the same reason as we do for `Thread`.

Thoughts?

Reply:
Oh, I really like this! It also addresses the problem of storing a pointer to the task somewhere else before running the thread (from a cursory look at the uses of `kthread_create`, it seems to be fairly common).

I'm not too concerned about having to call `get_task_struct`. Creating a thread is already an expensive operation anyway.